# Problem 26

https://projecteuler.net/problem=26

This is a fun problem! Knowing the decimal expansion of $\frac{1}{7} = 0.142857...$ is often a fun trick, and I often like to compute cycle lengths on public transport or on walks. Without further ado let's get started.

In [39]:
numTarget = 999

## Brute Force Solution

A first idea is to compute the reciprocal directly and inspect its decimal expansion. So let's just define the following method:

In [40]:
def cycleFinder(num: int) -> int:
    reciprocal = float(1/num)
    #...

But wait, this doesn't work! Python doesn't store these decimals to arbitrary precision, and we have no guarantee that the entire repeating cycle has been stored! 

Instead we have to try another approach. 

Instead, we can simulate the long division algorithm by hand. The key observation is that the decimal expansion begins to repeat precisely when a remainder repeats.

In [41]:
def cycleFinder(num: int) -> tuple[int, list[int], int]:
    digits = []
    seenRemainders = {}

    remainder = 1 % num
    position = 0
    cycleLength = 0

    while remainder not in seenRemainders and remainder != 0:
        seenRemainders[remainder] = position

        digit = (remainder * 10) // num
        digits.append(digit)

        remainder = (remainder * 10) % num
        position += 1

    if remainder == 0:
        cycleStart= position
        cycleLength = 0
    else:
        cycleStart = seenRemainders[remainder]
        cycleLength = position - cycleStart
    return cycleLength, digits, cycleStart

In [42]:
longestCycle = 0
bestNum = 0

for num in range(2, numTarget):
    length, digits, cycleStart = cycleFinder(num)

    if length > longestCycle:
        longestCycle = length 
        bestNum = num
        bestDigits = digits
        bestCycleStart = cycleStart

nonRepeating = ''.join(map(str, bestDigits[:bestCycleStart]))
repeating = ''.join(map(str, bestDigits[bestCycleStart:]))

print(f"1/{bestNum} = 0.{nonRepeating}({repeating})")

1/983 = 0.(0010172939979654120040691759918616480162767039674465920651068158697863682604272634791454730417090539165818921668362156663275686673448626653102746693794506612410986775178026449643947100712105798575788402848423194303153611393692777212614445574771108850457782299084435401831129196337741607324516785350966429298067141403865717192268565615462868769074262461851475076297049847405900305188199389623601220752797558494404883011190233977619532044760935910478128179043743641912512716174974567650050864699898270600203458799593082400813835198372329603255340793489318413021363173957273652085452695829094608341810783316378433367243133265513733468972533062054933875890132248219735503560528992878942014242115971515768056968463886063072227873855544252288911495422177009155645981688708036622583926754832146490335707019328585961342828077314343845371312309257375381485249237029501525940996948118006103763987792472024415055951169888097660223804679552390640895218718209562563580874872838250254323499491353)


And we get an answer of $983$.

For reference, its reciprocal has a period of 982! So we will not write it out fully.